pkg/kube/client/ssa/caching_unstructured_extractor.go

```diff
@@ -0,0 +1,261 @@
+// that it extracts the status subresource applied configuration.
+// Experimental!
+func (e *cachingUnstructuredExtractor) ExtractStatus(object *unstructured.Unstructured, fieldManager string) (*unstructured.Unstructured, error) {
+	return e.extractUnstructured(object, fieldManager, "status")
+}
+
+// getParserForGV fetches the *GVKParser for the given GVK.
+func (e *cachingUnstructuredExtractor) getParserForGV(ctx context.Context, gv schema.GroupVersion) (*GvkParser, error) {
+	data, err := e.dc.RESTClient().Get().
+		AbsPath("/openapi/v3").
+		Do(ctx).
+		Raw()
+
+	if err != nil {
+		return nil, err
+	}
+
+	discoMap := &handler3.OpenAPIV3Discovery{}
+	err = json.Unmarshal(data, discoMap)
+	if err != nil {
+		return nil, err
+	}
+	// parse discovery information
+	oapiPathsToGV := map[string]OpenAPIGroupVersion{}
+	for path, oapiGV := range discoMap.Paths {
+		parse, err := url.Parse(oapiGV.ServerRelativeURL)
+		if err != nil {
+			return nil, err
+		}
+		useClientPrefix := strings.HasPrefix(oapiGV.ServerRelativeURL, "/openapi/v3")
+		etag := parse.Query().Get("hash")
+		oapiPathsToGV[path] = newCustomOAPIGroupVersion(e.dc.RESTClient(), oapiGV, useClientPrefix, etag)
+	}
+
+	e.cache.mu.Lock()
+	defer e.cache.mu.Unlock()
+	// invalidate stale entries in cache with the fresh discovery data
+	for gvCached, cacheEntry := range e.cache.store {
+		path := gvRelativeAPIPath(gvCached)
+		if discoGV, ok := oapiPathsToGV[path]; !ok || discoGV.ETag() != cacheEntry.etag {
+			delete(e.cache.store, gvCached)
+		}
+	}
+
+	gvPath := gvRelativeAPIPath(gv)
+	oapiGV, ok := oapiPathsToGV[gvPath]
+	if !ok {
+		return nil, fmt.Errorf("cannot find GroupVersion %q in discovery", gvPath)
+	}
+
+	// check the cache after invalidating stale data
+	parserTuple, ok := e.cache.store[gv]
+	// generate new parser on cache miss, etag mismatch
+	// defensively cover the case where discovery does not return any ETag
+	// for GV, which normally should not happen
+	if !ok || parserTuple.etag != oapiGV.ETag() || oapiGV.ETag() == "" {
+		freshParser, err := newParserFromOpenAPIGroupVersion(ctx, oapiGV)
+		if err != nil {
+			return nil, err
+		}
+		e.cache.store[gv] = &GvkParserCacheEntry{
+			parser: freshParser,
+			etag:   oapiGV.ETag(),
+		}
+		return freshParser, nil
+	}
+	return parserTuple.parser, nil
+}
+
+// gvRelativeAPIPath constructs the OpenAPI path for the given GVK
+func gvRelativeAPIPath(gv schema.GroupVersion) string {
+	if gv.Group == "" {
+		return "api/" + gv.Version
+	}
+	return "apis/" + gv.String()
+}
+
+func newParserFromOpenAPIGroupVersion(ctx context.Context, oapiGV OpenAPIGroupVersion) (*GvkParser, error) {
+	s, err := oapiGV.Schema(ctx, "application/json")
+	if err != nil {
+		return nil, err
+	}
+	var oapi spec3.OpenAPI
+	if err := json.Unmarshal(s, &oapi); err != nil {
+		return nil, err
+	}
+
+	specs := map[string]*spec.Schema{}
+	for k, v := range oapi.Components.Schemas {
+		specs[k] = v
+	}
```

Comment:
after this, but before we build the gvk parser, we want to defensively double-check here that all openapi `$ref`s are part of specs

I made some similar thing that recursively traversed all linked schemas of a component; which looked like this:

```go
func ignoreEmptyStructDefaults(x *spec.Schema) {
	if x == nil {
		return
	}
	m, isStringMap := x.Default.(map[string]any)
	if isStringMap && len(m) == 0 {
		x.Default = nil
	}

	if x.AdditionalItems != nil {
		ignoreEmptyStructDefaults(x.AdditionalItems.Schema)
	}
	if x.AdditionalProperties != nil {
		ignoreEmptyStructDefaults(x.AdditionalProperties.Schema)
	}
	for i := range x.AllOf {
		ignoreEmptyStructDefaults(&x.AllOf[i])
	}
	for i := range x.AnyOf {
		ignoreEmptyStructDefaults(&x.AnyOf[i])
	}
	if x.Items != nil {
		ignoreEmptyStructDefaults(x.Items.Schema)
		for i := range x.Items.Schemas {
			ignoreEmptyStructDefaults(&x.Items.Schemas[i])
		}
	}
	ignoreEmptyStructDefaults(x.Not)
	for i := range x.OneOf {
		ignoreEmptyStructDefaults(&x.OneOf[i])
	}
	for k, s := range x.PatternProperties {
		ignoreEmptyStructDefaults(&s)
		x.PatternProperties[k] = s
	}
	for k, s := range x.Properties {
		ignoreEmptyStructDefaults(&s)
		x.Properties[k] = s
	}
}
```
it would be much better to actually use some of the "real" openapi walkers, something like this https://github.com/kubernetes/kube-openapi/blob/f7e401e7b4c2/pkg/schemamutation/walker.go#L55 + walkschema

and if you see a ref which isn't part of the specs, set some flag that this was invalid. there's no way to stop processing though in that code

Reply:
done, see https://github.com/turkenh/provider-kubernetes/pull/3/commits/8828f1e0418852b6e0c34c0229b69294dd9328fe